### PR TITLE
pyCA: 2.1 -> 4.5

### PR DESCRIPTION
**Pull request by pmiddend**
###### Description of changes

In light of ZHF, I noticed the package I maintain, pyCA, was failing, and updated it and fixed it at the same time (the tests were failing on all OSs because of sqlalchemy-2 being the standard in nixpkgs.python3Packages).

ZHF #230712

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pmiddend** [comment]
@ofborg build pyCA

**Comment by mweinelt** [comment]
It builds now, but since 3.0 the repo seems to have node dependencies specified, probably for some frontend build

https://github.com/opencast/pyCA/blob/v4.5/package.json

They could be built using `buildNpmPackage`, if they are needed.

**Comment by pmiddend** [comment]
Ahh, good that you mention this. I'll look into it and try to package it.

**Comment by pmiddend** [comment]
@mweinelt I added a build for the UI and "attached" the UI path to the `static_folder` argument for the Flask server. This way, you can just start the ui with `pyca ui` and it'll work (I've tested this).

File: pkgs/applications/video/pyca/default.nix [hunk 1]
```diff
@@ -1,25 +1,70 @@
...
+      repo = "pyCA";
+      rev = "v${version}";
+      sha256 = "sha256-cTkWkOmgxJZlddqaSYKva2wih4Mvsdrd7LD4NggxKQk=";
+    };
+
+    npmDepsHash = "sha256-0U+semrNWTkNu3uQQkiJKZT1hB0/IfkL84G7/oP8XYY=";
+
+    nativeBuildInputs = [ python ];
+
+    patches = [ ./add-package-json-version.patch ];
```
Comment (Mic92) [inline pkgs/applications/video/pyca/default.nix:33]: That version needs to be updated whenever we update the npm package version?
Maybe put a placeholder in the patch like `@version` and than use the substitute function in bash to replace it.

Comment (Mic92) [reply pkgs/applications/video/pyca/default.nix:33]: It's also possible to use `jq` and it's setpath function

Comment (pmiddend) [reply pkgs/applications/video/pyca/default.nix:33]: Uhh, I like the `jq` idea, I'll try that.

The version really is just to shut up whatever `buildNpm` does, it's complaining that a proper package needs a name and a version.

File: pkgs/applications/video/pyca/default.nix [hunk 2]
```diff
@@ -1,25 +1,70 @@
...
     sdnotify
+    psutil
     flask
+    prometheus-client
   ];
 
+  postPatch = ''
+    sed -i -e 's#static_folder=.*#static_folder="${frontend}/static")#' pyca/ui/__init__.py
+  '';
+
```
Comment (mweinelt) [inline pkgs/applications/video/pyca/default.nix:67]:
```suggestion

  passthru = {
    inherit frontend;
  };

```

Allows building `pyCA.frontend` to iterate on just the frontend.

Comment (pmiddend) [reply pkgs/applications/video/pyca/default.nix:70]: Didn't even know about that. Fixed!

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -5640,7 +5640,7 @@ with pkgs;
 
   waitron = callPackage ../tools/networking/waitron { };
 
-  pyCA = python3Packages.callPackage ../applications/video/pyca { };
+  pyCA = python3Packages.callPackage ../applications/video/pyca { inherit jq; };
```
Comment (mweinelt) [inline pkgs/top-level/all-packages.nix:5643]: That inherit seems unnecessary?

Comment (pmiddend) [reply pkgs/top-level/all-packages.nix:5643]: Yep, I thought so, too. But there is `python3Packages.jq` which will be used without this `inherit`

Comment (pmiddend) [reply pkgs/top-level/all-packages.nix:5643]: Okay, fixed.

**Review by mweinelt — APPROVED** [review]

**Pull request merged**